File: 프로그래머스/lv2/131704. 택배상자/택배상자.js
function solution(order) {
    order = order.reverse();
    const main = new Array(order.length).fill(0).map((_, i) => i+1).reverse(); // 오름차순 컨테이너
    let count = 0;
    let cur = order[order.length - 1];
    const sub = [];
    
    while(1) {
        // 메인에 있는경우
        if(!isEmpty(main) && main[main.length - 1] === cur ) {
            count++;
            order.pop();
            cur = order[order.length - 1];
            main.pop();
        }
        // 메인에 없는 경우
        else {
            // 서브에 있는 경우
            if(!isEmpty(sub) && sub[sub.length - 1] === cur) {
                count++;
                order.pop();
                cur = order[order.length - 1];
                sub.pop();
            }
            // 서브에 없는데 메인에 상자가 남아있는 경우
            else if(!isEmpty(main)) {
                sub.push(main.pop());
            }
            // 서브에도 없고 메인에도 상자가 없는 경우 
            else {
                break;
            }
        }
    }    
    
    return count;
}

const isEmpty = stack => {
    return stack.length === 0
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
from collections import deque

def solution(order):
    orders = deque(order)
    main_con = deque([i + 1 for i in range(len(order))])
    sub_con = []
    
    c = 0
    while(True):
        if len(orders) <= 0:
            break
        # 종료 조건 잊지말기
        cur_order = orders[0]
        # 종료 조건 1 - main이 비었는데 sub탑이 아닌경우
        if len(main_con) <= 0:
            if len(sub_con) <= 0:
                break
                
            if sub_con[-1] != cur_order:
                break
                
            else: # 일치하면 그거 뺌. 
                sub_con.pop()
                orders.popleft()
                c += 1
                continue
        
        cur_box = main_con[0]
        
        # main에 없으면
        if cur_box != cur_order:
            # sub에 있으면
            if len(sub_con) > 0 and sub_con[-1] == cur_order:
                c += 1
                orders.popleft()
                sub_con.pop()
            # sub에도 없으면
            else:
                sub_con.append(cur_box)
                main_con.popleft()
                
        # main에 있으면
        else:
            main_con.popleft()
            c += 1
            orders.popleft()
                
    return c
                
            
    
